### ISSUE TITLE
Improve discovery and description of JupyterLab commands

### ISSUE THREAD

**Issue by bollwyvl** [jupyterlab/jupyterlab#13962]
### Problem

JupyterLab commands are at once one of the most powerful, but also one of the most weakly-typed, and ultimately, least discoverable parts of the overall architecture.

Once a command id has actually been _discovered_ (by DOM, source code, or debugger inspection without source maps), it is then sometimes only possible to determine the appropriate `args` by fully tracing the source code from call site to ultimate implementation, as these are sometimes passed as `any` down into another part of the application. 

### Proposed Solution

- add `builder/metadata_schema.json#/properties/commands`
- for each `command`, _allow_ the definition of any of the static members of `ICommandOptions`, but _require_:
  - `#/describedBy/args`
    - as a JSON Schema, _require_ a `title` and `description`
  - `id`
    - while the `id` would ultimately be the _correct_ key for this object, having an arbitrary (but unique-per-package) string would allow for transparently shimming existing exported `CommandIds`
- move all core `CommandId`s and hard-coded strings to this location
  - import the `package.json` at runtime
    - in a `composite` build this... complicates things
      - each `packages/*/src` would need a `tsconfig.json` that `references: {path: ../}`
- use the above to also provide build-time typescript types for all args with [ajv](https://ajv.js.org/guide/typescript.html#utility-type-for-jtd-data-type) 

The payoff for this to be able to statically derive, outside of :

 - always-accurate generated static HTML documentation, which can be linked to and referenced
 - improved in-application documentation
   - Inspector/Contextual Help
   - discoverable forms a la #13951
     - this would work great with the sidebar command palette, but the modal one would be... not great
     - i guess it could launch _layers of blocking modals_ :nauseated_face:  

This technique could then be encouraged in extension-cookiecutter-ts, documentation, etc.

### References
- closer to reproducible client state #6013
- more impact from #13746
- a [certain extension](https://github.com/deathbeds/jyg) would _love_ to consume this data :blush:

**Comment by krassowski** [comment]
> Once a command id has actually been discovered (by DOM, source code, or debugger inspection without source maps),

For what its worth: https://jupyterlab.readthedocs.io/en/latest/user/commands.html#commands-list but this does not give us commands registered in extensions.

I do not understand _where_ you propose the actual JSON command to be defined. `package.json`?

There are more extensions that could benefit from having this information at runtime:
- https://github.com/jupyterlab/jupyterlab-plugin-playground
- https://github.com/jtpio/ipylab
- https://github.com/jupyterlab/ui-profiler

**Comment by krassowski** [comment]
An alternative wold be to discover this automatically by writing a TypeScript plugin with a pass-through transformer that would detect `addCommand` calls and dump the `args` type to declarations/JSON file. Thoughts?

**Comment by bollwyvl** [comment]
> `package.json`?

Yep. So, for example, this [CommandIds](https://github.com/jupyterlab/jupyterlab/blob/v4.0.0a33/packages/filebrowser-extension/src/index.ts#LL74C11-L74C21) would turn into an entry in [`package.json`](https://github.com/jupyterlab/jupyterlab/blob/v4.0.0a33/packages/filebrowser-extension/package.json#L65).

```json
{
  "name":  "name": "@jupyterlab/filebrowser-extension",
  // ...
  "jupyterlab": {
    "extension": true,
    "schemaDir": "schema",
    "commands": {
      // ...
      "showBrowser": {
        "id": "filebrowser:activate",
        "options": {
          "describedBy": {
            "args": {
              "description": "Open the file browser for the provided `path`.",
              "type": "object",
              "properties": {
                "path": {
                  "description": "the path to open",
                  "format": "uri",
                  "type": "string"
                }
              }
            }
          }
        },
      },
      // ...
    }
  }
}
```

Or, as in the alternative, it could go in [the plugin](https://github.com/jupyterlab/jupyterlab/blob/v4.0.0a33/packages/filebrowser-extension/schema/browser.json#LL2C2-L2C29) under a new key, `jupyter.lab.commands`.

Either way, the `CommandIds` would get exposed back in typescript with e.g.

```ts
import * as PACKAGE from "../package.json";
const Commands = PACKAGE.jupyterlab.commands;
```

And then installed, probably with a helper which could hoist `description` to `label` (or `caption`, or `usage`, or whatever) and translate it:

```ts
import { describeCommand } from `@jupyterlab/apputils`;
// ...
commands.addCommand(
  Commands.showBrowser.id, 
  describeCommand(
    translator,
    {
      ...Commands.showBrowser.options,
      activate: (args) => {...}
    }
  )
);
```

**Comment by bollwyvl** [comment]
> TypeScript plugin

Sure, that's a way to go, too. Got the same argument from the LSP folk who invented a _whole new metamodel_ for describing stuff. Cue xkcd link.

I'm still of the mind, for Jupyter, that metadata in an established, language-neutral format in a well-known (or at least describable) locations wins, especially when typescript can _natively_ import it, with decent typing. `grep` and `jq` (or literally any other language) could tear through all this stuff _without_ nodejs and typescript.

**Comment by bollwyvl** [comment]
> dump the `args` type to declarations/JSON file. 

Right, the issue is that today they _aren't_ described (basically all `any`), and sometimes just handed off to somewhere else. 

Writing schema-in-TS is of course possible, and could even be quite nice with authoring-type "magic" schema-to-ts by e.g. `ajv` (without actually _importing_ `ajv` everywhere).

**Comment by vidartf** [comment]
Should we alternatively make it easier for extension authors to expose their standard commands as tokens? I don't think that will be more work than speccing out the args in a schema.

**Comment by Mr-Ruben** [comment]
Here you can see the pain of 'command discovery' in order to edit the settings (cell toolbar).

[jupyterlab-how-to-insert-custom-elements-buttons-commands-in-the-cell-toolbar](https://stackoverflow.com/questions/78303136/jupyterlab-how-to-insert-custom-elements-buttons-commands-in-the-cell-tool/78303157#78303157)

If there is an easier way to get the list of available commands, I would really love to hear it.

**Comment by krassowski** [comment]
@Mr-Ruben today:
- For JupyterLab commands: https://jupyterlab.readthedocs.io/en/latest/user/commands.html#commands-list
- For extension commands you can list all commands programmatically  from https://github.com/jtpio/ipylab

But for the use case you are highlighting, the commands should really be a list/dropdown in the settings form.
